lib/server.js

```diff
@@ -1223,7 +1070,122 @@ Server.prototype._setupRequest = function _setupRequest(req, res) {
     if (self.name !== '') {
         res.setHeader('Server', self.name);
     }
-    res.version = self.router.versions[self.router.versions.length - 1];
+
+    // Request lifecycle events
+    // attach a listener for 'close' events, this will let us set
+    // a flag so that we can stop processing the request if the client closes
+    // the connection (or we lose the connection).
+    function onClose() {
+        res.removeListener('finish', onFinish);
+        res.removeListener('error', onError);
+        req._connectionState = 'close';
+
+        self._finishReqResCycle(req, res, new errors.RequestCloseError());
+    }
+    req.once('close', onClose);
+
+    // Response lifecycle events
+    function onFinish() {
+        req.removeListener('close', onClose);
+        res.removeListener('error', onError);
+        self._finishReqResCycle(req, res, res.err);
+    }
+    function onError(err) {
+        req.removeListener('close', onClose);
+        res.removeListener('finish', onFinish);
+        self._finishReqResCycle(req, res, err);
+    }
+    res.once('finish', onFinish);
+    res.once('error', onError);
+
+    // attach a listener for the response's 'redirect' event
+    res.on('redirect', function onRedirect(redirectLocation) {
+        self.emit('redirect', redirectLocation);
+    });
+};
+
+/**
+ * Maintaining the end of the request-response cycle:
+ *  - emitting after event
+ *  - updating inflight requests metrics
+ * Check if the response is finished, and if not, wait for it before firing the
+ * response object.
+ *
+ * @private
+ * @memberof Server
+ * @instance
+ * @function _finishReqResCycle
+ * @param {Object} req - the request object
+ * @param {Object} res - the response object
+ * @param {Object} [err] - a possible error as a result of failed route matching
+ * or failed execution of the handler array.
+ * @returns {undefined} no return value
+ */
+Server.prototype._finishReqResCycle = function _finishReqResCycle(
+    req,
+    res,
+    err
+) {
+    var self = this;
+    var route = req.route; // can be undefined when 404 or error
+
+    // res.finished is set by node core's response object, when
+    // res.end() completes. if the request was closed by the client, then emit
+    // regardless of res status.
+
+    // after event has signature of function(req, res, route, err) {...}
+    if (!res.finished && !req.closed()) {
+        res.once('finish', function resFinished() {
```

Comment:
Since `finishReqResCycle()` doesn't get triggered until finish event fires, is this scenario possible anymore?